File: 1891-cutting-ribbons/1891-cutting-ribbons.py
class Solution:
    def maxLength(self, ribbons: List[int], k: int) -> int:
        

        l = 1 ; r = max(ribbons)  
        while l <=r: 

            m = (l+r) >>1 

            cnt = 0 
            for rib in ribbons: 
                cnt += rib//m 



            if cnt >= k: 
                l = m +1  
            else: 
                r = m-1 

        return  r    

File: 1891-cutting-ribbons/1891-cutting-ribbons.java
class Solution {
    public int maxLength(int[] ribbons, int k) {
        
        int l = 100000;
        int r = 0 ; 
        for(var rib:ribbons){
            l = 0 ;
            r = Math.max(r, rib) ; 
        }
        
        
        while(l<r){
            int m = (l + r + 1) / 2    ;        
            int tot= 0 ; 
            if(m !=0){
                for(var rib:ribbons){
                tot += rib /m ; 
            }
            }
          
            
            if( tot >= k){
                l = m  ;
            }else{
                r = m -1 ;
            }
            
        }
        return r ; 
        
    }
}